presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/HistoryTable.java

```diff
@@ -0,0 +1,94 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.metadata.InternalTable;
+import io.prestosql.spi.connector.ColumnMetadata;
+import io.prestosql.spi.connector.ConnectorPageSource;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.ConnectorTableMetadata;
+import io.prestosql.spi.connector.ConnectorTransactionHandle;
+import io.prestosql.spi.connector.FixedPageSource;
+import io.prestosql.spi.connector.SystemTable;
+import io.prestosql.spi.predicate.TupleDomain;
+import org.apache.iceberg.HistoryEntry;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.util.SnapshotUtil;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP;
+
+public class HistoryTable
+        implements SystemTable
+{
+    private final IcebergTableHandle tableHandle;
+    private final Table icebergTable;
+
+    private static final List<ColumnMetadata> COLUMNS = ImmutableList.of(
+            new ColumnMetadata("made_current_at", TIMESTAMP),
+            new ColumnMetadata("snapshot_id", BIGINT),
+            new ColumnMetadata("parent_id", BIGINT),
+            new ColumnMetadata("is_current_ancestor", BOOLEAN));
+
+    public HistoryTable(IcebergTableHandle tableHandle, Table icebergTable)
+    {
+        this.tableHandle = tableHandle;
+        this.icebergTable = icebergTable;
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return Distribution.SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return new ConnectorTableMetadata(tableHandle.getSchemaTableName(), COLUMNS);
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        Map<Long, Snapshot> snapshots = new HashMap<>();
+        for (Snapshot snap : icebergTable.snapshots()) {
+            snapshots.put(snap.snapshotId(), snap);
+        }
+        Set<Long> ancestorIds = ImmutableSet.copyOf(SnapshotUtil.currentAncestors(icebergTable));
+
+        InternalTable.Builder historyTable = InternalTable.builder(COLUMNS);
+        for (HistoryEntry historyEntry : icebergTable.history()) {
+            long snapshotId = historyEntry.snapshotId();
+            Snapshot snapshot = snapshots.get(snapshotId);
```

Comment:
Can't we use `icebergTable.snapshot(snapshotId)` here instead of building Map<Long, Snapshot> snapshots?

Reply:
Good point. I didn't see that.